frontend-react/src/components/RSErrorBoundary.tsx

```diff
@@ -0,0 +1,97 @@
+import React, { ErrorInfo, ReactNode, Suspense } from "react";
+
+import {
+    ERROR_UI_DEFAULT,
+    ErrorName,
+    ErrorUI,
+    isRSError,
+    RSError,
+} from "../utils/RSError";
+import { ErrorComponent } from "../pages/error/ErrorComponent";
+
+import Spinner from "./Spinner";
+
+interface ErrorBoundaryProps {
+    children?: ReactNode;
+}
+interface ErrorBoundaryState {
+    hasError: boolean;
+    // Undefined until an error is thrown
+    code?: ErrorName;
+    type?: ErrorUI;
+}
+/** Used to define default state values on render */
+const initState: ErrorBoundaryState = {
+    hasError: false,
+};
+/** Wrap components with this error boundary to catch errors thrown */
+export default class RSErrorBoundary extends React.Component<
+    ErrorBoundaryProps,
+    ErrorBoundaryState
+> {
+    constructor(props: any) {
+        super(props);
+        this.state = initState;
+    }
+    /** Handles parsing error and setting up state accordingly */
+    static getDerivedStateFromError(error: RSError): ErrorBoundaryState {
+        /* We throw a lot of errors, so refactoring seems out of scope, but the system built relies on
+         * RSError functionality to render unique error pages */
+        const useLegacy = !isRSError(error);
+        if (useLegacy) {
+            console.warn(
+                "Please work to migrate all non RSError throws to use an RSError-extending error type."
+            );
+        }
+        return {
+            hasError: true,
+            code: useLegacy ? ErrorName.NON_RS_ERROR : error.code,
+            type: useLegacy ? ERROR_UI_DEFAULT : error.displayAs,
+        };
+    }
+    /** Any developer logging needed (i.e. log the error in console, push to
+     * analytics, etc.) */
+    componentDidCatch(error: RSError, errorInfo: ErrorInfo) {
+        console.error(error, errorInfo);
+    }
+    /** Renders the right error page for the right error type, OR the wrapped
+     * component if no error is thrown */
+    render() {
+        if (this.state.hasError) {
+            return (
+                <ErrorComponent code={this.state.code} ui={this.state.type} />
```

Comment:
[wild thought] what if, rather than an `ErrorComponent` that could either represent an in page message or an entire page of content, we maintain those two separate options and switch off of using them here based on the state. So bear with me - in all other cases where ErrorComponent is used in this PR the ui is hardcoded, this is the only place where it is dynamic. Instead of making the ui a prop to a higher level component here, what if we added an `errorComponent` state param that can be derived based on the error type?

What's here definitely works, but I wonder if that would alleviate some confusion when implementing error handling for specific situations, allowing users to specifically use a page or message component rather than something higher level.

Or maybe this is a bad idea???

Reply:
I'm re-considering this because technically the only difference here _is_ the component. A "message" should ideally not replace the whole component, but just prepend the error banner. I'm gonna make some changes to this and get a re-review in a bit 👍🏼